pkg/api/dashboard.go

```diff
@@ -746,6 +746,67 @@ func (hs *HTTPServer) GetDashboardVersion(c *models.ReqContext) response.Respons
 	return response.JSON(http.StatusOK, dashVersionMeta)
 }
 
+// swagger:route POST /dashboards/validate dashboards validateDashboard
+//
+// Validates a dashboard JSON against the schema
+//
+// Produces:
+// - application/json
+//
+// Responses:
+// 200: validateDashboardResponse
+// 401: unauthorisedError
+// 403: forbiddenError
+// 500: internalServerError
+func (hs *HTTPServer) ValidateDashboard(c *models.ReqContext) response.Response {
+	cmd := models.ValidateDashboardCommand{}
+
+	if err := web.Bind(c.Req, &cmd); err != nil {
+		return response.Error(http.StatusBadRequest, "bad request data", err)
+	}
+
+	cm := hs.Coremodels.Dashboard()
+
+	// POST api receives dashboard as a string of json (so line numbers for errors stay consistent),
+	// but we need to parse the schema version out of it
+	dashboardBytes := []byte(cmd.Dashboard)
+	dashboardJson, err := simplejson.NewJson(dashboardBytes)
+	if err != nil {
+		return response.Error(http.StatusBadRequest, "unable to parse dashboard", err)
+	}
+
+	schemaVersion, err := dashboardJson.Get("schemaVersion").Int()
+
+	isValid := false
+	validationMessage := ""
+
+	// Only try to validate if the schemaVersion is at least the handoff version
+	// (the minimum schemaVersion against which the dashboard schema is known to
+	// work), or if schemaVersion is absent (which will happen once the Thema
+	// schema becomes canonical).
+	if err != nil || schemaVersion >= dashboard.HandoffSchemaVersion {
+		v, _ := cuectx.JSONtoCUE("dashboard.json", dashboardBytes)
+		_, validationErr := cm.CurrentSchema().Validate(v)
+
+		if validationErr == nil {
+			isValid = true
+		} else {
+			validationMessage = validationErr.Error()
+		}
+	} else {
+		validationMessage = "invalid schema version"
+	}
+
+	respData := &ValidateDashboardResponse{
+		IsValid: isValid,
+		Message: validationMessage,
+	}
+
+	// We respond with StatusOK even when the dashboard failed validation we were
+	// able to "task failed successfully" on an invalid dashboard
+	return response.JSON(http.StatusOK, respData)
```

Comment:
I'm now thinking that maybe the API shouldn't respond with 200 when a dashboard fails validation, mainly to make it easier to track metrics for validation failures from the existing logging.\

I'm not sure what the status codes should be? Should the 'invalid schema version' status code be different to the 'failed schema validation' error? (keeping in mind the response message is the theme validation error string when it fails validation)

thoughts @sakjur ?

Reply:
Hm, yeah, that makes sense, [`http.StatusUnprocessableEntity`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) seems appropriate?